ndd_slurm.py

```diff
@@ -184,42 +186,79 @@ def get_slaves(args):
         raise ValueError('one of -D or -d must be specified')
 
 
-def run_slave(args):
-    procs = []
+def start_slave_tar(args, procs):
+    read_fd, write_fd = os.pipe()
+    procs.append(init_process(['tar', '-xC', args.o], read_fd=read_fd))
+    return write_fd
+
+
+def start_slave_pack(args, procs):
+    read_fd, write_fd = os.pipe()
     if args.r:
-        read_fd, write_fd = os.pipe()
-        procs.append(init_process(['tar', '-xC', args.o], read_fd=read_fd))
+        out_fd = start_slave_tar(args, procs)
     else:
+        out_fd = os.open(args.o, os.O_CREAT | os.O_WRONLY)
+    procs.append(init_process(['pigz', '-d'], read_fd=read_fd,
+                              write_fd=out_fd))
+    return write_fd
+
+
+def run_slave(args):
+    procs = []
+    if not (args.r or args.z):
         write_fd = None
+    if args.z:
+        write_fd = start_slave_pack(args, procs)
+    elif args.r:
+        write_fd = start_slave_tar(args, procs)
     if args.H:
-        slave = args.c
-        slaves = args.S.split(',')
-        cmd = get_ssh_slave_ndd_cmd(args, slave, slaves, write_fd)
+        cmd = get_ssh_slave_ndd_cmd(args, write_fd)
     else:
         cmd = get_slave_ndd_cmd(args, write_fd)
-    procs.append(init_process(cmd))
+    procs.append(init_process(cmd, write_fd=write_fd))
     return wait(procs)
 
 
-def run_master(args):
-    procs = []
+def start_master_tar(args, procs):
+    read_fd, write_fd = os.pipe()
+    procs.append(init_process(['tar', '-c', args.i],
+                              write_fd=write_fd))
+    return read_fd
+
+
+def start_master_pack(args, procs):
+    read_fd, write_fd = os.pipe()
     if args.r:
-        read_fd, write_fd = os.pipe()
-        procs.append(init_process(['tar', '-c', args.i], write_fd=write_fd))
+        in_fd = start_master_tar(args, procs)
     else:
+        in_fd = os.open(args.i, os.O_RDONLY)
+    procs.append(init_process(['pigz', '--fast'], read_fd=in_fd,
+                              write_fd=write_fd))
+    return read_fd
+
+
+def run_master(args):
+    procs = []
+    if not (args.r or args.z):
         read_fd = None
-    cmds = [get_master_ndd_cmd(args, read_fd)]
-    cmds += get_ssh_cmds(args) if args.H else [get_srun_cmd(args)]
-    procs.extend(map(init_process, cmds))
+    if args.z:
+        read_fd = start_master_pack(args, procs)
```

Comment:
It would be better to move all the logic with `args.z` and `args.r` into `start_master_pack`.